### ISSUE TITLE
`~`-operator for inversion does not work as expected on torch.Tensor.byte in v1.2.0

### ISSUE THREAD

**Issue by benjamin9555** [pytorch/pytorch#24180]
## 🐛 Bug

The command  `~(torch.tensor([1, 0]).byte())` outputs `tensor([254, 255], dtype=torch.uint8)`

## Expected behavior

output should be `tensor([0, 1], dtype=torch.uint8)`

## Environment

PyTorch version: 1.2.0.dev20190805
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX TITAN X
GPU 1: GeForce GTX TITAN X
GPU 2: GeForce GTX TITAN X
GPU 3: GeForce GTX TITAN X
GPU 4: GeForce GTX TITAN X
GPU 5: GeForce GTX TITAN X
GPU 6: GeForce GTX TITAN X
GPU 7: GeForce GTX TITAN X

Nvidia driver version: 418.40.04
cuDNN version: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.2.0.dev20190805
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl_fft                   1.0.12           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch-nightly           1.2.0.dev20190805 py3.7_cuda10.0.130_cudnn7.6.2_0    pytorch

**Comment by ezyang** [comment]
cc @izdeby but IIRC this is expected behavior, and if you want boolean-style inversion you should use `logical_not` (but this PR did not make it into 1.2.0 https://github.com/pytorch/pytorch/pull/23839 )

**Comment by izdeby** [comment]
Logical_not is not yet implemented but there is a [PR](https://github.com/pytorch/pytorch/pull/23839) for it already.
In case of bool tensors, use `bitwise_not`
`torch.bitwise_not(torch.tensor([True, False]))`

**Comment by ezyang** [comment]
On a bool tensor, `~` operator should DTRT, right?

**Comment by izdeby** [comment]
Yes, `~` and `bitwise_not` are equal operators. 
```
>>> torch.bitwise_not(torch.tensor([True, False]))
tensor([False,  True])

>>> ~torch.tensor([True, False])
tensor([False,  True])
```

**Comment by gchanan** [comment]
It's "expected" although we shouldn't have made this change in this release.

The mental model of masking in this release is "bool is preferred, byte is deprecated", but for inversion the model is "bool works, byte doesn't", which is inconsistent. The alternative is some complicated documentation of the ~ operator, but that seems justified here.

Since we've already made this change though, I'd recommend changing your indexing to boolean (assuming that's why you are using uint8 in the first place).

**Comment by ezyang** [comment]
I'm closing this for now as not a bug. Please reopen if you disagree.
